### ISSUE TITLE
OFT grid_sample lowering: tile-padded constants and channel-expanded broadcast cause DRAM OOM (32x blowup)

### ISSUE THREAD

**Issue by ctr-lelanchelianTT** [tenstorrent/tt-mlir#8321]
### Describe the bug

The StableHLO → TTIR/TTNN lowering for `torch.nn.functional.grid_sample` (as produced by torch_xla / tt-xla) materializes huge intermediate tensors that exhaust DRAM during execution. The OOM presents as a tt-metal `bank_manager.cpp` `TT_FATAL`, but it is a tt-mlir lowering issue — the same `ttnn.repeat` op runs fine in isolation on the same device.

This is the same bug reported in tt-xla [#4244](https://github.com/tenstorrent/tt-xla/issues/4244) (current main) and previously [#3419](https://github.com/tenstorrent/tt-xla/issues/3419) (Feb branch).

### Failing op (model-level)

```
TT_FATAL @ tt_metal/impl/allocator/bank_manager.cpp:462: false
Out of Memory: Not enough space to allocate 1449713664 B DRAM buffer across 12 banks,
where each bank needs to store 120809472 B, but bank size is 1071821792 B
(allocated: 970619392 B, free: 101202400 B, largest free block: 101202400 B)

backtrace:
  ttnn::prim::RepeatDeviceOperation::create_output_tensors
  ttnn::prim::repeat
  ttnn::operations::data_movement::detail::repeat_upper_dims_rm
  ttnn::repeat
  tt::runtime::ttnn::operations::data_movement::run(RepeatOp, ...)
```

### Why this is a tt-mlir issue, not a tt-metal issue

I built a single-op tt-metal sanity test (file: `tests/ttnn/unit_tests/operations/data_movement/test_repeat_oft_oom.py`) with four progressively-larger setups:

| # | Setup | Result |
|---|---|---|
| 1 | `ttnn.repeat` alone, input `(1,1,7,25281,2)` × `(1,256,1,1,1)` | **PASSED** — 1.45 GB allocation succeeds on a clean device |
| 2 | `to_layout(TILE)` → `reshape` → `repeat` | **PASSED** |
| 3 | Pre-allocate `ttnn.full((1,256,7,25281,1), INT32, TILE)` (the `main_const_eval_2` from EmitPy), then run upstream + repeat | **FAILED** — `5,805,965,312 B` OOM (exact match to #3419) |
| 4 | Pre-allocate two pressure tensors (INT32 and FP32 of `(1,256,7,25281,*)`), then run upstream + repeat | **FAILED** — `1,449,713,664 B` with `allocated: 970,032,000 B` (exact match to the OFT model OOM in #4244) |

Full log: https://gist.github.com/ctr-lelanchelianTT/263502bc23b9117a3b7e7d9c6e54b552

So `ttnn.repeat` on its own is fine. The OOM is driven entirely by **upstream tile-padded intermediates that the lowering chooses to materialize**.

### IR-level smoking guns

From the StableHLO that tt-mlir receives for a single `grid_sample(input=(1,256,28,28), grid=(1,7,25281,2))` (full IR in `oft_grid_sample/irs/shlo_compiler_cleaned_*.mlir`, generated via `examples/pytorch/codegen/python/grid_sample.py` in tt-xla):

```mlir
%c = stablehlo.constant dense<0> : tensor<1x256x7x25281x1xi64>             // smoking gun #1
%5 = stablehlo.broadcast_in_dim %arg0, dims = [0, 2, 3, 4]
        : (tensor<1x7x25281x2xf32>) -> tensor<1x256x7x25281x2xf32>          // smoking gun #2
```

**#1**: The `i64` all-zero constant of shape `(1, 256, 7, 25281, 1)` is materialized as a TTNN tensor. After lowering to INT32 in TILE layout, the last two dims must be tile-padded to multiples of 32. `25281 → 25312` and `1 → 32`, giving an effective shape of `(1, 256, 7, 25312, 32)` and a DRAM footprint of `256 * 7 * 25312 * 32 * 4 B ≈ 5.8 GB`. That single allocation is enough to cause #3419's exact OOM.

**#2**: The grid (shape `(1, 7, 25281, 2)`, ~1.4 MB) is broadcast across the channel axis to `(1, 256, 7, 25281, 2)`, which then lowers to a `ttnn.repeat` with output ~1.45 GB. Combined with #1 already on device, this is the OOM seen in #4244.

### Reproduction (using existing artifacts)

```bash
# In tt-xla:
python examples/pytorch/codegen/python/grid_sample.py        # produces oft_grid_sample/
cd oft_grid_sample
./run                                                         # triggers the OOM via ttnn

# Inspect the input IR that tt-mlir sees:
less oft_grid_sample/irs/shlo_compiler_cleaned_*.mlir         # has the two smoking guns

# In tt-metal (proves it isn't a tt-metal bug):
pytest -svv tests/ttnn/unit_tests/operations/data_movement/test_repeat_oft_oom.py
```

### Suggested fix direction

Two related changes in `lib/Conversion/StableHLOToTTIR/`:

1. **Constant folding for all-zero/all-N constants**: A `stablehlo.constant dense<K>` of any shape should not be materialized as a fully-allocated TTNN tensor, especially when the result has tile-padding waste. It should be lowered to `ttnn.full` with the original (unpadded) logical shape, and ideally kept symbolic until the consumer forces materialization. The current path produces `(1, 256, 7, 25281, 1) i64 → INT32 TILE` which is the worst case for tile padding (last_dim=1 → 32× waste).

2. **Avoid materializing `broadcast_in_dim` before a gather**: When a `stablehlo.broadcast_in_dim` is consumed by a `stablehlo.gather` (or by indexing ops that decompose into gather), the broadcast should be fused/elided rather than lowering to `ttnn.repeat`. Materializing the channel-expanded grid is the dominant cost in grid_sample lowering.

Either fix in isolation should resolve the OOM for the OFT model; both are desirable for general correctness/efficiency.

### Related

- tt-xla #4244 — model-level report on current main
- tt-xla #3419 — same bug from the Feb branch, ~4× larger allocation; same root cause

**Comment by ctr-lelanchelianTT** [comment]
### Implementation pointer for the fix

After tracing the lowering, here's the picture:

**Where the splat constants currently land in TTIR:**
`oft_grid_sample/irs/ttir_*.mlir` shows constants like:
```mlir
%0 = "ttir.constant"() <{value = dense<0> : tensor<1x256x7x25281x1xi64>}> : () -> tensor<1x256x7x25281x1xi64>
```
…materialized as a full `ttir.constant` carrying the dense value, even though it is a single-value splat.

**Pattern that emits `ttir.constant`:**
[`lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:1154-1224`](https://github.com/tenstorrent/tt-mlir/blob/main/lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp#L1154-L1224) — `StableHLOToTTIRConstantOpConversionPattern::matchAndRewrite`.

**TTIR → TTNN constant lowering:**
[`lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp:1076-1142`](https://github.com/tenstorrent/tt-mlir/blob/main/lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp#L1076-L1142) — `ConstantOpConversionPattern::matchAndRewrite`. Crucially, it picks layout from `layoutAttr.isTiled()`, so when the upstream layout decision picked TILE the constant gets tile-padded.

**Available named creation ops** (already defined, with folders):
- `TTIR_FullOp` — `include/ttmlir/Dialect/TTIR/IR/TTIROps.td:5109` (`shape`, `fill_value`)
- `TTIR_ZerosOp` — `…:5234` (`shape`)
- `TTIR_OnesOp` — `…:5273` (`shape`)

### Suggested two-step fix

**Step 1 (small, low-risk): splat-aware lowering in `StableHLOToTTIRConstantOpConversionPattern`.**

Before falling through to `ttir.constant`, detect splat constants and emit a named creation op:

```cpp
// pseudocode inside matchAndRewrite, after newValueAttr is built
if (auto denseAttr = dyn_cast<DenseElementsAttr>(newValueAttr);
    denseAttr && denseAttr.isSplat()) {
  Attribute splat = denseAttr.getSplatValue<Attribute>();

  if (auto intSplat = dyn_cast<IntegerAttr>(splat);
      intSplat && intSplat.getValue().isZero()) {
    rewriter.replaceOpWithNewOp<ttir::ZerosOp>(srcOp, outputType,
        rewriter.getDenseI32ArrayAttr(toI32Vec(outputType.getShape())));
    return success();
  }
  if (auto fpSplat = dyn_cast<FloatAttr>(splat);
      fpSplat && fpSplat.getValue().isZero()) {
    rewriter.replaceOpWithNewOp<ttir::ZerosOp>(srcOp, outputType,
        rewriter.getDenseI32ArrayAttr(toI32Vec(outputType.getShape())));
    return success();
  }
  // Optional: ones, then ttir.full for arbitrary scalar splats.
  rewriter.replaceOpWithNewOp<ttir::FullOp>(srcOp, outputType, splat);
  return success();
}
// else: existing ttir.constant path
```

This alone:
- removes the multi-MB `dense<…>` blobs from TTIR (today the i64 zero blob for `(1,256,7,25281,1)` gets carried through several passes)
- exposes the operations to the existing folders on `TTIR_FullOp` / `TTIR_ZerosOp`
- gives downstream layout/L1-fallback analyses a clearer signal that the value is splat (so they can pick a row-major / on-device-fill path instead of paying tile-pad cost)

It does **not by itself** force a row-major lowering, so it may not flip the OOM on its own. But it is a prerequisite for step 2 and is independently a cleanup.

**Step 2 (the actual OOM fix): layout-aware lowering for splat creation ops with small `last_dim`.**

In `lib/Dialect/TTNN/Analysis/LegalTensorLayoutAnalysis.cpp` (or the closest equivalent on the constant/full path), when the producer is a `ttir::ZerosOp`/`OnesOp`/`FullOp` and the result type has `last_dim < 32` (or `last_two_dims_pad_ratio > N`), prefer `Layout::RowMajor`. Tile-pad cost for `(…, 1)` is 32× and that's the dominant cost in this model.

Concretely the trigger we want to catch:

```
shape (1, 256, 7, 25281, 1) i64 + TILE  ->  pad to (1, 256, 7, 25312, 32) -> 5.8 GB
shape (1, 256, 7, 25281, 1) i64 + ROW_MAJOR -> 181 MB           (32x reduction)
```

A heuristic in the layout-selection pass to pick row-major when `tile_padded_volume / actual_volume > 8` would resolve the OFT case without changing behavior for shapes where TILE is appropriate.

### Reproduction artifacts

- StableHLO: `oft_grid_sample/irs/shlo_compiler_cleaned_*.mlir` (see line 4 for the splat constant, line 19 for the broadcast that becomes `ttnn.repeat`)
- TTIR after lowering: `oft_grid_sample/irs/ttir_*.mlir` (lines 7–16 are all `ttir.constant` splats that should be named creation ops)
- TTNN: `oft_grid_sample/irs/ttnn_*.mlir`
- Generated EmitPy: `oft_grid_sample/main.py:39-49` (`main_const_eval_2`) shows the `ttnn.full(..., layout=TILE)` that explodes
- tt-metal isolated repro: see prior gist

If/when env-activation issues are sorted on my side, I am happy to draft Step 1 as a self-contained PR (with a lit test under `test/ttmlir/Conversion/StableHLOToTTIR/`).
